qiskit/chemistry/algorithms/ground_state_solvers/orbital_optimization_vqe.py

```diff
@@ -0,0 +1,663 @@
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A ground state calculation employing the Orbital-Optimized VQE (OOVQE) algorithm.
+"""
+
+from typing import Optional, List, Union, Tuple
+import logging
+import copy
+import numpy as np
+from scipy.linalg import expm
+from qiskit.aqua import AquaError
+from qiskit.aqua.algorithms import VQE
+from qiskit.aqua.operators import LegacyBaseOperator
+
+from .ground_state_eigensolver import GroundStateEigensolver
+from .minimum_eigensolver_factories import MinimumEigensolverFactory
+from ...fermionic_operator import FermionicOperator
+from ...bosonic_operator import BosonicOperator
+from ...drivers.base_driver import BaseDriver
+from ...transformations.fermionic_transformation import FermionicTransformation
+from ...results.electronic_structure_result import ElectronicStructureResult
+from ...qmolecule import QMolecule
+
+logger = logging.getLogger(__name__)
+
+
+class OrbitalOptimizationVQE(GroundStateEigensolver):
+    r""" A ground state calculation employing the OOVQE algorithm.
+    The Variational Quantum Eigensolver (VQE) algorithm enhanced with the Orbital Optimization (OO).
+    The core of the approach resides in the optimization of orbitals through the
+    AO-to-MO coefficients matrix C. In the usual VQE, the latter remains constant throughout
+    the simulation. Here, its elements are modified according to C=Ce^(-kappa) where kappa is
+    an anti-hermitian matrix. This transformation preserves the spectrum but modifies the
+    amplitudes of the ground state of given operator such that in the end a given ansatz
+    can be closest to that ground state, producing larger overlap and lower eigenvalue than
+    conventional VQE.  Kappa is parametrized and optimized inside the OOVQE in the same way as
+    the gate angles. Therefore, at each step of OOVQE the coefficient matrix C is modified and
+    the operator is recomputed, unlike usual VQE where operator remains constant.
+    Iterative OO refers to optimization in two steps, first the wavefunction and then the
+    orbitals. It allows for faster optimization as the operator is not recomputed when
+    wavefunction is optimized. It is recommended to use the iterative method on real device/qasm
+    simulator with noise to facilitate the convergence of the classical optimizer.
+    For more details of this method refer to: https://aip.scitation.org/doi/10.1063/1.5141835
+    """
+
+    def __init__(self,
+                 transformation: FermionicTransformation,
+                 solver_factory: MinimumEigensolverFactory,
```

Comment:
This parameter is called 'solver' in the other ground state solvers. This one, algorithm the GroundStateSolver ought to be able to be configured more directly with a VQE solver since as with a heuristic var form it can all be built out and just given directly here - no need to create the solver later

Reply:
changed to `solver` and allows `MinimumEigenSolver` as well as factory. However, raises exception in case it isn't UCCSD (same for return of factory).